show.md

```diff
@@ -0,0 +1,207 @@
+
+## Background and motivation
+
+Most CAD software has the concept of *construction geometry* which is geometry used for constructing other geometry but which is not itself rendered. We currently don't have explicit support for this in KittyCAD, but the engine does make some choices about which geometry is rendered. In KCL, non-rendered geometry might take other forms too, e.g., geometry in functions or in imported modules. It's also likely to be used in many places, e.g., creating a single instance of an object and then using a pattern to create multiple instances, where we don't want to render the original instance.
+
+One particular question to answer is how geometry should work with imported modules (which I envisage to be the way that KCL supports assemblies). To recap the current and [proposed](https://github.com/KittyCAD/kcl-experiments/pull/12) syntax:
+
+- modules are declared in a separate file
+- exported functions of a file can be imported using `import foo from 'bar.kcl'`
+- the module itself (i.e., the geometry defined at the top-level of the file) can be imported using `import 'bar.kcl'` (creates a single object bound to the variable `bar`, which should be able to be used as an assembly)
+
+Currently, objects are effectively rendered by virtue of the side effects of execution (issuing API calls to the engine). Therefore, exactly how all the above might work is a bit vague. I'll spell some of it out here, focussing on the user-facing syntax. More of the fundamentals are discussed in the [foundations design doc](foundations.md), including side-effects.
+
+
+## Proposal: implicit `show`
+
+The basics of the proposal is that an object that is defined at the top level is rendered by the engine unless it is passed into a function, used in a pipeline, or assigned into a variable (or used in some other way we add to KCL in the future). Using the result of a function or a variable at the top level will render it. E.g.,
+
+```
+// Result of function call, rendered
+makeSphere()
+
+// Result of pipeline, rendered (nothing intermediate in the pipeline is rendered)
+startSketchOn(...)
+  |> startPath()
+  |> ...
+  |> close()
+  |> extrude()
+
+// As above, but assigned into a variable, nothing is rendered
+object = startSketchOn(...)
+  |> startPath()
+  |> ...
+  |> close()
+  |> extrude()
+
+// Variable is used, result of pipeline is rendered
+object
+```
+
+Exactly how this is implemented with respect to the engine is left to the [foundations design doc](foundations.md), however, whether something is rendered or not must (I think) be explicit in the API calls.
+
+When a function is declared, nothing is rendered (if an entity escapes usage into the top-level of the function, then there should be an error or warning). When a function is called, the result returned by the function follows the usage rules as described above.
```

Comment:
So, I think mechanical refactoring of code into functions is a good guideline, but not an absolute principle. Consider Java for example, in general you can copy and paste into a function, but it's not 100% straightforward because you might reference variables which are out of scope (the IDE can make these into arguments, so mechanical, but not trivial). Then consider Rust, the same thing is mostly true but actually due to the borrowing rules there are times when it is not possible to do this, so not only not trivial but also not possible (and not even possible to predict without a very complete model of the borrow checker). So, in KCL we can basically copy and paste, but we are going to need to identify arguments, and we are also going to have to insert a `return` for where something might other wise be rendered. That seems reasonable to me.

Partly because (and addressing your second para) there is not a straightforward way to render geometry in functions, consider `fn unitCircle(centre) { circle(centre, radius = 2) }`. There is no way to render that. You might say pick a strawman argument (e.g., the origin), but what if radius were the argument? The default is much less obvious and there is no easy way to know whether a default argument is a good idea or not (what if there are two points, picking both to be the origin would probably not work - it might for an elipse, but not for a line). But there's a bigger problem - the circle is a 2d object, how do you choose a plane/surface to render it on?

Anyway, afaict, there is not good way to render geometry in a function and we shouldn't try to do so by default, i.e., an error is the right thing to do. We probably should have some support in the UI for this (e.g., the user selects the geometry in the function and the UI can suggest values for missing variables and the user can manipulate them for debugging).